File: tools/test-core-flow.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Playwright test for S06-T02:
- index.html intro sequence
- intranet.html save/load & slider backtrack
"""

import json
import sys
import time
from playwright.sync_api import sync_playwright

BASE_URL = "http://localhost:8888"
RESULT_FILE = "tools/test-core-flow-result.txt"


def log(msg):
    print(msg)
    with open(RESULT_FILE, "a", encoding="utf-8") as f:
        f.write(msg + "\n")


def main():
    with open(RESULT_FILE, "w", encoding="utf-8") as f:
        f.write("S06-T02 Core Flow Test Started\n")

    errors = []
    ac1_pass = False
    ac2_pass = False
    ac3_pass = False

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(viewport={"width": 1280, "height": 800})
        page = context.new_page()

        def on_error(err):
            msg = str(err)
            errors.append(msg)
            log(f"[JS ERROR] {msg}")

        page.on("pageerror", on_error)
        page.on("console", lambda msg: log(f"[CONSOLE {msg.type}] {msg.text}") if msg.type == "error" else None)

        # ============================================================
        # AC-1: index.html intro -> navigate to intranet.html
        # ============================================================
        log("\n=== AC-1: index.html intro sequence ===")
        try:
            page.goto(f"{BASE_URL}/index.html")
            page.wait_for_load_state("networkidle")

            # Step 0 -> next
            page.locator("#next-btn").wait_for(state="visible", timeout=10000)
            page.locator("#next-btn").click()

            # Step 1 -> click camera
            page.locator("#item-camera").wait_for(state="visible", timeout=10000)
            page.locator("#item-camera").click()

            # Wait for photo-1 active and click
            page.wait_for_selector("#photo-slot-1.active", timeout=10000)
            page.locator("#photo-slot-1").click()

            # Modal open -> close
            page.wait_for_selector("#photo-modal.show", timeout=10000)
            page.locator("#modal-close-btn").click()

            # Next through steps 3, 4, 5
            for _ in range(3):
                page.locator("#next-btn").wait_for(state="visible", timeout=10000)
                page.locator("#next-btn").click()

            # Search
            page.wait_for_selector("#search-box-1.visible", timeout=10000)
            page.locator("#search-input").fill("云处村")
            page.locator("#search-btn").click()
            page.wait_for_selector("#search-result .search-result-item", timeout=10000)

            # Wait for hidden text then next
            page.wait_for_selector("#search-result .hidden-text", timeout=15000)
            page.locator("#next-btn").wait_for(state="visible", timeout=10000)
            page.locator("#next-btn").click()

            # Email toast
            page.wait_for_selector("#email-toast.show", timeout=10000)
            page.locator("#email-toast-open").click()

            # Inline email -> confirm
            page.wait_for_selector("#email-inline.show", timeout=10000)
            page.locator("#email-confirm-btn").click()

            # Photo 2
            page.wait_for_selector("#photo-slot-2.active", timeout=10000)
            page.locator("#photo-slot-2").click()
            page.wait_for_selector("#photo-modal.show", timeout=10000)
            page.locator("#modal-close-btn").click()

            # Final next -> end
            page.locator("#next-btn").wait_for(state="visible", timeout=10000)
            page.locator("#next-btn").click()

            # Wait for explore button
            page.wait_for_selector("#explore-btn", timeout=10000)

            # Navigate to intranet
            page.goto(f"{BASE_URL}/intranet.html")
            page.wait_for_load_state("networkidle")

            if errors:
                log(f"AC-1: JS errors observed: {errors}")
            else:
                log("AC-1: Intro sequence completed without JS errors.")
            ac1_pass = True
        except Exception as e:
            log(f"AC-1 FAIL: {e}")
            ac1_pass = False

        # ============================================================
        # AC-2: intranet save / load
        # ============================================================
        log("\n=== AC-2: intranet save/load ===")
        try:
            # Ensure clean state
            page.evaluate("localStorage.removeItem('yunchu_save')")
            page.reload()
            page.wait_for_load_state("networkidle")

            # Wait for toast and start chat
            page.wait_for_selector("#msg-toast.show", timeout=10000)
            page.locator("#msg-toast").click()
            page.wait_for_selector("#panel-chat.active", timeout=10000)

            def click_choice(text):
                btn = page.locator("#choice-row").get_by_text(text, exact=False)
                btn.wait_for(state="visible", timeout=20000)
                btn.click()

            # Advance through Act1 choices
            click_choice("你们到底在说什么")
            click_choice("你怎么知道")
            click_choice("拯救")
            click_choice("什么目标")
            click_choice("被困在时间里的人")

            # Back to home -> map -> continue map chat
            page.locator("#panel-chat .chat-back-btn").click()
            page.wait_for_selector("#panel-home.active", timeout=10000)
            page.locator("#card-map").click()
            page.wait_for_selector("#panel-map.active", timeout=10000)
            page.locator("#map-continue-btn").click()
            page.wait_for_selector("#panel-chat.active", timeout=10000)

            # Advance map chat to >= 8 (starts at 8, choice pushes to 9/10/11)
            click_choice("时间循环")
            click_choice("什么孩子")

            # Verify localStorage
            save_raw = page.evaluate("() => localStorage.getItem('yunchu_save')")
            save = json.loads(save_raw or "{}")
            log(f"localStorage yunchu_save: {json.dumps(save, ensure_ascii=False)}")
            cp = save.get("chatProgress", 0)
            if cp >= 8:
                log(f"AC-2: chatProgress = {cp} (>= 8) saved correctly.")
            else:
                raise AssertionError(f"chatProgress too low: {cp}")

            # Refresh and resume
            page.reload()
            page.wait_for_load_state("networkidle")
            page.locator("#card-chat").click()
            page.wait_for_selector("#panel-chat.active", timeout=10000)

            chat_text = page.locator("#chat-messages").inner_text()
            if "云处村是一个畲族村落" not in chat_text:
                raise AssertionError("Chat did not resume after refresh")
            log("AC-2: Chat resumed correctly after refresh.")
            ac2_pass = True
        except Exception as e:
            log(f"AC-2 FAIL: {e}")
            ac2_pass = False

        # ============================================================
        # AC-3: slider backtrack
        # ============================================================
        log("\n=== AC-3: slider backtrack ===")
        try:
            # Go back to home so slider is visible
            page.locator("#panel-chat .chat-back-btn").click()
            page.wait_for_selector("#panel-home.active", timeout=10000)
            page.wait_for_selector("#progress-bar-wrap.visible", timeout=10000)

            track = page.locator("#progress-slider-track")
            thumb = page.locator("#progress-slider-thumb")
            track.wait_for(state="visible", timeout=10000)
            thumb.wait_for(state="visible", timeout=10000)

            tb = track.bounding_box()
            hb = thumb.bounding_box()
            if not tb or not hb:
                raise RuntimeError("Could not get slider bounding boxes")

            # Drag thumb to ~50% of track height (Act1 when maxActIdx=2)
            start_x = hb["x"] + hb["width"] / 2
            start_y = hb["y"] + hb["height"] / 2
            target_x = start_x
            target_y = tb["y"] + tb["height"] * 0.5

            page.mouse.move(start_x, start_y)
            page.mouse.down()
            page.mouse.move(target_x, target_y)
            page.mouse.up()

            # Wait for log panel to become active
            page.wait_for_selector("#panel-log.active", timeout=10000)

            log_text = page.locator("#log-messages").inner_text()
            if "你终于来了" not in log_text:
                raise AssertionError("Slider backtrack did not restore chapter 1 content")
            log("AC-3: Slider drag switched to panel-log and restored chapter 1.")
            ac3_pass = True
        except Exception as e:
            log(f"AC-3 FAIL: {e}")
            ac3_pass = False

        browser.close()

    # ============================================================
    # Summary
    # ============================================================
    log("\n=== Summary ===")
    log(f"AC-1: {'PASS' if ac1_pass else 'FAIL'}")
    log(f"AC-2: {'PASS' if ac2_pass else 'FAIL'}")
    log(f"AC-3: {'PASS' if ac3_pass else 'FAIL'}")

    if not (ac1_pass and ac2_pass and ac3_pass):
        sys.exit(1)


if __name__ == "__main__":
    main()

File: tools/test-core-flow.js
const { chromium } = require('playwright');

const BASE_URL = 'http://localhost:8080';

(async () => {
  const browser = await chromium.launch({ headless: true });
  const page = await browser.newPage();
  const consoleErrors = [];

  page.on('console', msg => {
    if (msg.type() === 'error') {
      consoleErrors.push(msg.text());
      console.log('❌ Console error:', msg.text());
    }
  });

  page.on('pageerror', err => {
    consoleErrors.push(err.message);
    console.log('❌ Page error:', err.message);
  });

  // ===== Test 1: index.html intro flow =====
  console.log('=== Test 1: index.html intro flow ===');
  await page.goto(`${BASE_URL}/index.html`);
  await page.waitForTimeout(1000);

  await page.waitForSelector('#next-btn');
  await page.click('#next-btn');
  await page.waitForTimeout(1000);

  await page.waitForSelector('#item-camera.visible');
  await page.click('#item-camera');
  await page.waitForTimeout(1500);

  await page.waitForSelector('#photo-slot-1.active');
  await page.click('#photo-slot-1');
  await page.waitForTimeout(600);

  await page.click('#modal-close-btn');
  await page.waitForTimeout(1000);

  for (let i = 0; i < 3; i++) {
    await page.waitForSelector('#next-btn');
    await page.click('#next-btn');
    await page.waitForTimeout(1000);
  }

  await page.fill('#search-input', '云处村');
  await page.click('#search-btn');
  await page.waitForTimeout(2500);

  await page.waitForSelector('#next-btn');
  await page.click('#next-btn');
  await page.waitForTimeout(1000);

  await page.waitForSelector('#email-toast.show');
  await page.click('#email-toast-open');
  await page.waitForTimeout(800);

  await page.click('#email-confirm-btn');
  await page.waitForTimeout(1000);

  await page.waitForSelector('#photo-slot-2.active');
  await page.click('#photo-slot-2');
  await page.waitForTimeout(600);

  await page.click('#modal-close-btn');
  await page.waitForTimeout(1000);

  await page.waitForSelector('#next-btn');
  await page.click('#next-btn');
  await page.waitForTimeout(2500);

  const exploreBtn = await page.locator('#explore-btn').count();
  console.log(`Explore button present: ${exploreBtn > 0}`);

  if (exploreBtn > 0) {
    await page.click('#explore-btn');
    await page.waitForURL('**/yunchu-gov.html', { timeout: 8000 });
    console.log('✅ Navigated to yunchu-gov.html');
  }

  console.log(`Test 1 console errors: ${consoleErrors.length}`);

  // ===== Test 2: intranet.html save/load =====
  console.log('\n=== Test 2: intranet.html save/load ===');
  await page.goto(`${BASE_URL}/intranet.html`);
  await page.waitForTimeout(2000);

  await page.evaluate(() => {
    localStorage.setItem('yunchu_save', JSON.stringify({
      chatProgress: 8,
      maxProgress: 8,
      radioStarted: false,
      foerStarted: false,
      shadowStarted: false,
      researchStarted: false,
      revealStarted: false
    }));
    location.reload();
  });

  await page.waitForTimeout(3000);

  const saveData = await page.evaluate(() => {
    return JSON.parse(localStorage.getItem('yunchu_save') || '{}');
  });
  console.log('Save data after reload:', saveData);
  console.log(`chatProgress restored: ${saveData.chatProgress === 8}`);

  // ===== Test 3: slider backtrack =====
  console.log('\n=== Test 3: progress slider backtrack ===');
  const sliderTrackExists = await page.locator('#progress-slider-track').count() > 0;
  const sliderThumbExists = await page.locator('#progress-slider-thumb').count() > 0;
  console.log(`Slider track exists: ${sliderTrackExists}`);
  console.log(`Slider thumb exists: ${sliderThumbExists}`);

  let panelBeforeSlider = null;
  let panelAfterSlider = null;

  if (sliderTrackExists && sliderThumbExists) {
    panelBeforeSlider = await page.evaluate(() => {
      const active = document.querySelector('.page-panel.active');
      return active ? active.id : null;
    });
    console.log(`Active panel before slider: ${panelBeforeSlider}`);

    const track = await page.locator('#progress-slider-track').boundingBox();
    if (track) {
      await page.mouse.click(track.x + track.width / 2, track.y + track.height - 20);
      await page.waitForTimeout(1200);
      console.log('Clicked on slider track near bottom');
    }

    panelAfterSlider = await page.evaluate(() => {
      const active = document.querySelector('.page-panel.active');
      return active ? active.id : null;
    });
    console.log(`Active panel after slider: ${panelAfterSlider}`);
  }

  const choiceBtns = await page.locator('.choice-btn').count();
  console.log(`Choice buttons available: ${choiceBtns}`);

  await page.screenshot({ path: '/tmp/test3-intranet-slider.png', fullPage: true });

  // Summary
  const summary = {
    test1_introFlow: {
      navigatedToGov: exploreBtn > 0,
      consoleErrors: consoleErrors.length,
      errorMessages: consoleErrors
    },
    test2_localStorage: {
      chatProgressRestored: saveData.chatProgress === 8,
      saveData
    },
    test3_slider: {
      sliderTrackExists,
      sliderThumbExists,
      panelChanged: panelBeforeSlider !== panelAfterSlider,
      panelBefore: panelBeforeSlider,
      panelAfter: panelAfterSlider,
      choicesAvailable: choiceBtns
    }
  };

  console.log('\n=== Summary ===');
  console.log(JSON.stringify(summary, null, 2));

  // Write result file
  const fs = require('fs');
  fs.writeFileSync('/tmp/test-core-flow-result.json', JSON.stringify(summary, null, 2));
  console.log('\nResult saved to /tmp/test-core-flow-result.json');

  await browser.close();
})();
